### ISSUE TITLE
Better CI/CD or Full CI/CD

### ISSUE THREAD

**Issue by jot2re** [TokenScript/attestation#244]
We need to construct a more complete CI/CD flow during PRs. In particular when it comes to integration, both with authenticator.js and attestation.id.

**Comment by weiwu-zhang** [comment]
@abcfy2 do you know the hardhat testing flow?

This is actually not trivial - we might do a small project for getting this right, proper CI/CD setup between attestation.id, authenticator.js, attestation.jar and the smart contracts take a bit of effort to figure out.

**Comment by jot2re** [comment]
In relation to this it should also be mentioned that there are some issues with the hardhat tests in attestation.jar, see issue #254 
As part of this CI/CD task these should also be reenable.

**Comment by foxgem** [comment]
> there are some issues with the hardhat tests

yes, that's why I suggest we get our dependencies graph first, ;).

**Comment by jot2re** [comment]
It looks really good! Thanks @foxgem! 
I think the only thing we might also want to add is the smart contract hard hat testing when attestation.jar has been updated. See issue #254.

**Comment by oleggrib** [comment]
currently attestation JS code located under https://github.com/TokenScript/attestation/tree/main/src/main/javascript/crypto
to test it I do next steps:
1. in the _root_ folder  ```gradle test --rerun-tasks``` to regenerate test output files
2. in the _main/src/main/javascript/crypto_ folder I run ```npm run testjest``` , it read output files from _gradle test_ and tests agains them.

So we can run grable test and npm test one by one or find another solution and generate JAVA output before each NPM test

another note - TN includes _attestation npm_ module and use it to decode magicLink, to test attestation, make proof etc., so ideally TN should be updated when _attestation npm_ changed

**Comment by foxgem** [comment]
## Phase#1

Goals: 

- attestation github clean up
- attestation <> attestation.id CI integration

### Action Plan

1. Attestation `main` branch
    1. java component (java-main-p1)
        1. build - test - grype
    2. js component (js-main-p1)
        1. build - test
    3. contract component (contract-main-p1)
        1. build - test
2. Attestation `stage` branch
    1. java component
        1. java-p1 - prerelease - attestation.id/backend-main-p1
        2. js-p1 - prerelease - attestation.id/frontend-main-p1
3. attestation.id `main` branch
    1. backend (backend-main-p1)
        1. build - test
    2. frontend (frontend-main-p1)
        1. build - test
4. attestation `main` enhancement
     1. java-main-p1: trigger js-main-p1 and contract-main-p1
     2. js-main-p1: security scanning
     3. contract-main-p1: security scanning

**Comment by foxgem** [comment]
after discussing with @jot2re , the output of `grype` will be like:

success

```
% grype attestation-0.3.18-all.jar 
 ✔ Vulnerability DB        [updated]
New version of grype is available: 0.40.0
 ✔ Indexed attestation-0.3.18-all.jar 
 ✔ Cataloged packages      [23 packages]
 ✔ Scanned image           [0 vulnerabilities]
```

failure

```
% grype build/libs/backend-0.1-all.jar 
 ✔ Vulnerability DB        [no update available]
New version of grype is available: 0.40.0
 ✔ Indexed build/libs/backend-0.1-all.jar 
 ✔ Cataloged packages      [83 packages]
 ✔ Scanned image           [25 vulnerabilities]

NAME                         INSTALLED     FIXED-IN      TYPE          VULNERABILITY        SEVERITY 
netty-codec-http             4.1.73.Final  4.1.77.Final  java-archive  GHSA-269q-hmxg-m83q  Medium    
netty-reactive-streams       2.0.5                       java-archive  CVE-2019-20445       Critical  
netty-reactive-streams       2.0.5                       java-archive  CVE-2021-21290       Medium    
netty-reactive-streams       2.0.5                       java-archive  CVE-2021-37136       High      
netty-reactive-streams-http  2.0.5                       java-archive  CVE-2015-2156        High
```

so, the `grype` task is to find if there is **"0 vulnerabilities"** in the output. cc @abcfy2

**Comment by abcfy2** [comment]
I've committed `java-p1` CI, which will run gradle build & test & grype check after `push` / `pull_request` to `main` branch.

Sample: https://github.com/TokenScript/attestation/actions/runs/2532698892

**Comment by abcfy2** [comment]
`js-p1` CI added. When `push`/`pull_request` code changes to `main` branch in `src/main/javascript/crypto/` directory will trigger this CI.

Sample: https://github.com/TokenScript/attestation/actions/runs/2536158552

**Comment by abcfy2** [comment]
`contract-p1` CI addes.

**Comment by abcfy2** [comment]
`stage` branch workflow finished.

Result sample:

java CI: https://github.com/TokenScript/attestation/actions/runs/2596014131

js CI: https://github.com/TokenScript/attestation/actions/runs/2595982268

**Comment by foxgem** [comment]
## Phase#2

Goals: including TN project

Length: 1~2 weeks

### Action Plan

1. TN `main` branch (tn-p1)
     - build - test
2. if the `stage` branch changed in `tokenscript/attestation` or `attestation.id`
     - pre-release-of-dependencies - call tn p1

**Comment by abcfy2** [comment]
Phase#2 done.

Sample: https://github.com/TokenScript/attestation/actions/runs/2648616646

**Comment by foxgem** [comment]
@abcfy2 as we discussed, please enhance the CI with E2E testing.
